src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+-- | > Subset A B
+-- | refers to identifiers @A@ and @B@. The function will perform a lookup in the symbol table, make sure the objects exist and are of desired types -- in this case, 'Set' -- and throws an error otherwise.
+passReferencess :: SubObjects -> SubStmt -> SubObjects
+passReferencess e (ApplyP (Predicate (PredicateConst t) args pos)) = e { subObjs = (toConstr t args) : subObjs e }
+passReferencess e _ = e -- Ignore all other statements
 
 -- | Similar to 'toObj'
-toConstr :: SubType -> [String] -> SubObj
-toConstr NoIntersectT [a, b] = LC $ NoIntersect a b
-toConstr IntersectT [a, b] = LC $ Intersect a b
-toConstr PointInT [a, b] = LC $ PointIn a b
-toConstr PointNotInT [a, b] = LC $ PointNotIn a b
-toConstr SubsetT [a, b] = LC $ Subset a b
-toConstr NoSubsetT [a, b] = LC $ NoSubset a b
-toConstr t os = error ("toConstr: incorrect arguments to " ++ show t ++ " "++ show os)
-
--- helper function that checks the symbol table before inserting, making sure there is no duplicated declarations
+toConstr :: String -> [PredArg] -> SubObj
+toConstr p vl = LC $ SubConstrConst p vl
+
+-- | helper function that checks the symbol table before inserting, making sure there is no duplicated declarations
 checkAndInsert s t m = case M.lookup s m of
     Nothing -> M.insert s t m
     _ -> error ("Duplicated symbol: " ++ s)
--- helper function that checks the existence and type of an object
-checkNameAndTyp m (s, t) = case M.lookup s m of
-    Nothing -> error ("Undefined symbol: " ++ s)
-    Just t' -> if t == t' then s
-            else error ("Type of " ++ s ++ " is incorrect. Expecting " ++ show t ++ " , but have " ++ show t')
 
--- | `subSeparate` aplits a list of Substance objects into declared objects and constaints on these objects
+-- | helper function that checks the existence and type of an object
+checkNameAndTyp :: (Show a, Eq a) => M.Map String a -> (String, a) -> String
+checkNameAndTyp m (s, t) = 
+                case M.lookup s m of
+                Nothing -> error ("Undefined symbol: " ++ s)
+                Just t' -> if t == t'
+                           then s
+                           else error ("Type of " ++ s ++ " is incorrect. Expecting " ++ show t ++ " , but have " ++ show t')
+
+-- | `subSeparate` splits a list of Substance objects into declared objects and constaints on these objects
 subSeparate :: [SubObj] -> SubObjDiv
 subSeparate = foldr separate ([], [])
-            where separate line (decls, constrs) =
-                           case line of
-                           (LD x) -> (x : decls, constrs)
-                           (LC x) -> (decls, x : constrs)
+              where separate line (decls, constrs) =
+                             case line of
+                             (LD x) -> (x : decls, constrs)
+                             (LC x) -> (decls, x : constrs)
 
+loadObjects :: SubProg -> VarEnv -> SubObjects
+loadObjects p subEnv = let env1 = foldl (passDecls subEnv) initObjs p
```

Comment:
misleading name -- [SubObjs] is not an environment

Reply:
I see you just did a big refactor of Nimo's code here, but I'm sure it could be made much simpler still :-)
for example:
- the subEnv is only used once (in declaredNames)
- checkAndInsert and checkNameAndTyp are not used anywhere in the codebase
- not sure that Style really needs us to do this `SubObj` conversion, since it's mostly putting the same info into a different container

Reply:
@hypotext : If you come up with a better interface for connecting Substance and Style while you implement Style let me know and I'll do the Substance work